### ISSUE TITLE
no-inner-declarations errors inside ArrowFunctionExpression

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#1892]
```js
// test.js
foo(() => {
    function bar() {
        return 1;
    }

    bar();
});
```

```js
// config.json
{
    "rules": {
        "no-inner-declarations": [2, "both"]
    },
    "ecmaFeatures": {
        "arrowFunctions": true
    }
}
```

```sh
$ eslint --reset --no-eslintrc -c config.json test.js

test.js
  3:4  error  Move function declaration to program root  no-inner-declarations

✖ 1 problem (1 error, 0 warnings)
```

Since `bar` is declared inside the body of an `ArrowFunctionExpression`, it should be considered valid.

**Comment by jrvidal** [comment]
Oops, I started working without asking. I can still submit a PR for #1893 if necessary.

Also, this problem could reappear for `class` methods, if they introduce a new AST node type.

**Comment by nzakas** [comment]
The fix has already been merged.

In the future, just add a comment saying "working on this" and when you expect to have it done. That way, others will know and let you finish. :)

**Comment by btmills** [comment]
@jrvidal I should've explicitly said I was fixing this one. Please do go ahead with the PR for #1893. Good point about class methods; we'll need to watch out for that once Espree gets classes, though for now estree/estree has methods as FunctionExpression nodes.
